File: voxform-go/internal/handlers/public.go
package handlers

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"

	"github.com/go-chi/chi/v5"

	"github.com/voxform/api/internal/db"
	"github.com/voxform/api/internal/models"
	"github.com/voxform/api/internal/queue"
)

type PublicHandler struct {
	db          *db.DB
	queue       *queue.Queue
	storagePath string
	storageURL  string
}

func NewPublicHandler(d *db.DB, q *queue.Queue, storagePath, storageURL string) *PublicHandler {
	return &PublicHandler{db: d, queue: q, storagePath: storagePath, storageURL: storageURL}
}

// GetBySlug returns a public survey with its questions.
func (h *PublicHandler) GetBySlug(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	survey, err := models.GetSurveyBySlug(h.db.DB, slug)
	if err != nil {
		NotFound(w, "survey not found or not active")
		return
	}

	questions, _ := models.ListQuestions(h.db.DB, survey.ID)

	OK(w, map[string]any{
		"id":          survey.ID,
		"title":       survey.Title,
		"description": survey.Description,
		"slug":        survey.Slug,
		"config":      survey.Config,
		"questions":   questions,
	})
}

// StartSession creates an anonymous respondent session for a public survey.
func (h *PublicHandler) StartSession(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	survey, err := models.GetSurveyBySlug(h.db.DB, slug)
	if err != nil {
		NotFound(w, "survey not found or not active")
		return
	}

	id := newID()
	h.db.Exec(`
		INSERT INTO sessions (id, survey_id, status, sync_source, started_at, device_info, created_at, updated_at)
		VALUES (?, ?, 'IN_PROGRESS', 'ONLINE', NOW(), '{}', NOW(), NOW())
	`, id, survey.ID)

	OK(w, map[string]any{"sessionId": id, "surveyId": survey.ID})
}

// SubmitResponse records a single question answer for an anonymous session.
func (h *PublicHandler) SubmitResponse(w http.ResponseWriter, r *http.Request) {
	sessionID := chi.URLParam(r, "id")

	var count int
	h.db.DB.Get(&count, `SELECT COUNT(*) FROM sessions WHERE id = ? AND status = 'IN_PROGRESS'`, sessionID)
	if count == 0 {
		NotFound(w, "session not found or already completed")
		return
	}

	var body struct {
		QuestionID       string   `json:"questionId"`
		Type             string   `json:"type"`
		TextValue        *string  `json:"textValue"`
		AudioDurationSec *float64 `json:"audioDurationSec"`
	}
	if err := Decode(r, &body); err != nil || body.QuestionID == "" || body.Type == "" {
		BadRequest(w, "questionId and type are required")
		return
	}

	id := newID()
	h.db.Exec(`
		INSERT INTO responses (id, session_id, question_id, type, text_value, audio_duration_sec, status, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, 'SUBMITTED', NOW(), NOW())
	`, id, sessionID, body.QuestionID, body.Type, body.TextValue, body.AudioDurationSec)

	Created(w, map[string]any{"responseId": id})
}

// CompleteSession marks an anonymous session as completed.
func (h *PublicHandler) CompleteSession(w http.ResponseWriter, r *http.Request) {
	sessionID := chi.URLParam(r, "id")
	h.db.Exec(`
		UPDATE sessions SET status = 'COMPLETED', completed_at = NOW(), updated_at = NOW()
		WHERE id = ? AND status = 'IN_PROGRESS'
	`, sessionID)
	OK(w, map[string]any{"sessionId": sessionID, "status": "COMPLETED"})
}

// CreateAudioSlot registers an upload intent for a public response.
func (h *PublicHandler) CreateAudioSlot(w http.ResponseWriter, r *http.Request) {
	var body struct {
		ResponseID string `json:"responseId"`
		MimeType   string `json:"mimeType"`
	}
	if err := Decode(r, &body); err != nil || body.ResponseID == "" {
		BadRequest(w, "responseId is required")
		return
	}

	ext := ".webm"
	if body.MimeType == "audio/wav" || body.MimeType == "audio/wave" {
		ext = ".wav"
	}

	uploadID := randHex(16)
	filename := uploadID + ext
	destPath := filepath.Join(h.storagePath, "audio", "raw", filename)
	if err := os.MkdirAll(filepath.Dir(destPath), 0755); err != nil {
		InternalError(w, err)
		return
	}

	var count int
	h.db.DB.Get(&count, `SELECT COUNT(*) FROM audio_jobs WHERE response_id = ?`, body.ResponseID)
	if count == 0 {
		h.db.Exec(`
			INSERT INTO audio_jobs (id, response_id, status, wav_path, provider, attempt_count, created_at, updated_at)
			VALUES (?, ?, 'UPLOADING', ?, 'GROQ', 0, NOW(), NOW())
		`, newID(), body.ResponseID, destPath)
	} else {
		h.db.Exec(`UPDATE audio_jobs SET status = 'UPLOADING', wav_path = ?, updated_at = NOW() WHERE response_id = ?`, destPath, body.ResponseID)
	}

	OK(w, map[string]any{
		"uploadId":  uploadID,
		"uploadUrl": fmt.Sprintf("%s/audio/raw/%s", h.storageURL, filename),
	})
}

// UploadAudio receives a multipart audio file for a public response and enqueues processing.
func (h *PublicHandler) UploadAudio(w http.ResponseWriter, r *http.Request) {
	uploadID := chi.URLParam(r, "uploadId")

	r.ParseMultipartForm(200 << 20)

	file, header, err := r.FormFile("file")
	if err != nil {
		BadRequest(w, "no file in request")
		return
	}
	defer file.Close()

	responseID := r.FormValue("responseId")
	if responseID == "" {
		BadRequest(w, "responseId is required")
		return
	}

	ext := filepath.Ext(header.Filename)
	if ext == "" {
		ext = ".webm"
	}
	if header.Header.Get("Content-Type") == "audio/wav" {
		ext = ".wav"
	}

	destPath := filepath.Join(h.storagePath, "audio", "raw", uploadID+ext)
	if err := os.MkdirAll(filepath.Dir(destPath), 0755); err != nil {
		InternalError(w, err)
		return
	}

	dst, err := os.Create(destPath)
	if err != nil {
		InternalError(w, err)
		return
	}
	defer dst.Close()

	written, err := io.Copy(dst, file)
	if err != nil {
		InternalError(w, err)
		return
	}

	h.db.Exec(`UPDATE audio_jobs SET wav_path = ?, status = 'QC_PENDING', updated_at = NOW() WHERE response_id = ?`, destPath, responseID)

	h.queue.Enqueue("audio:process", map[string]any{"responseId": responseID})

	Created(w, map[string]any{"bytes": written})
}

File: voxform-fastapi/routers/public.py
"""Public routes (no auth) — mirrors internal/handlers/public.go."""

import os

from fastapi import APIRouter, Depends, File, Form, HTTPException, Request, UploadFile
from databases import Database

import config
from auth_utils import new_id, rand_hex
from deps import get_db
from routers.audio import _do_upload

router = APIRouter(tags=["public"])


def _ok(data):
    return {"success": True, "data": data}


def _err(status: int, msg: str):
    raise HTTPException(status_code=status, detail={"success": False, "message": msg})


@router.get("/public/surveys/{slug}")
async def get_by_slug(slug: str, db: Database = Depends(get_db)):
    survey = await db.fetch_one(
        "SELECT * FROM surveys WHERE slug = :slug AND status = 'ACTIVE'", {"slug": slug}
    )
    if not survey:
        _err(404, "survey not found or not active")

    questions = await db.fetch_all(
        "SELECT * FROM questions WHERE survey_id = :sid ORDER BY order_index ASC",
        {"sid": survey["id"]},
    )
    return _ok({
        "id": survey["id"], "title": survey["title"],
        "description": survey["description"], "slug": survey["slug"],
        "config": survey["config"],
        "questions": [dict(q) for q in questions],
    })


@router.post("/public/surveys/{slug}/session")
async def start_session(slug: str, db: Database = Depends(get_db)):
    survey = await db.fetch_one(
        "SELECT id FROM surveys WHERE slug = :slug AND status = 'ACTIVE'", {"slug": slug}
    )
    if not survey:
        _err(404, "survey not found")

    sid = new_id()
    await db.execute(
        "INSERT INTO sessions (id, survey_id, status, sync_source, device_info) "
        "VALUES (:id, :survey, 'IN_PROGRESS', 'ONLINE', '{}')",
        {"id": sid, "survey": survey["id"]},
    )
    return _ok({"sessionId": sid, "surveyId": survey["id"]})


@router.post("/public/sessions/{session_id}/response", status_code=201)
async def submit_response(session_id: str, request: Request, db: Database = Depends(get_db)):
    body = await request.json()
    session = await db.fetch_one("SELECT id FROM sessions WHERE id = :id", {"id": session_id})
    if not session:
        _err(404, "session not found")

    rid = new_id()
    await db.execute(
        "INSERT INTO responses (id, session_id, question_id, type, text_value, "
        "audio_duration_sec, status) VALUES (:id, :sid, :qid, :type, :tv, :dur, 'SUBMITTED')",
        {
            "id": rid,
            "sid": session_id,
            "qid": body.get("questionId"),
            "type": body.get("type") or "SHORT_TEXT",
            "tv": body.get("textValue"),
            "dur": body.get("audioDurationSec"),
        },
    )
    return _ok({"responseId": rid})


@router.patch("/public/sessions/{session_id}/complete")
async def complete_session(session_id: str, db: Database = Depends(get_db)):
    session = await db.fetch_one("SELECT id FROM sessions WHERE id = :id", {"id": session_id})
    if not session:
        _err(404, "session not found")
    await db.execute(
        "UPDATE sessions SET status = 'COMPLETED', completed_at = NOW(), updated_at = NOW() WHERE id = :id",
        {"id": session_id},
    )
    return _ok({"sessionId": session_id, "status": "COMPLETED"})


@router.post("/public/images/upload/{response_id}", status_code=201)
async def upload_image(
    response_id: str,
    file: UploadFile = File(...),
    db: Database = Depends(get_db),
):
    session = await db.fetch_one(
        "SELECT id FROM sessions WHERE id = (SELECT session_id FROM responses WHERE id = :rid LIMIT 1)",
        {"rid": response_id},
    )
    if not session:
        _err(404, "response not found")

    ext = os.path.splitext(file.filename or "")[1].lower() or ".jpg"
    if ext not in (".jpg", ".jpeg", ".png", ".gif", ".webp", ".heic"):
        ext = ".jpg"

    img_dir = os.path.join(config.STORAGE_PATH, "images")
    os.makedirs(img_dir, exist_ok=True)
    save_path = os.path.join(img_dir, f"{response_id}{ext}")

    content = await file.read(50 * 1024 * 1024)  # 50 MB limit
    with open(save_path, "wb") as f:
        f.write(content)

    public_url = f"{config.STORAGE_URL}/images/{response_id}{ext}"
    await db.execute(
        "UPDATE responses SET text_value = :url, updated_at = NOW() WHERE id = :id",
        {"url": public_url, "id": response_id},
    )
    return _ok({"url": public_url, "bytes": len(content)})


@router.post("/public/audio/slot")
async def create_audio_slot(request: Request, db: Database = Depends(get_db)):
    body = await request.json()
    response_id = body.get("responseId") or ""
    mime_type = body.get("mimeType") or ""

    upload_id = rand_hex(8)
    ext = ".wav" if "wav" in mime_type else ".webm"

    raw_dir = os.path.join(config.STORAGE_PATH, "audio", "raw")
    os.makedirs(raw_dir, exist_ok=True)
    upload_path = os.path.join(raw_dir, f"{upload_id}{ext}")
    upload_url = f"{config.STORAGE_URL}/audio/raw/{upload_id}{ext}"

    existing = await db.fetch_one(
        "SELECT id FROM audio_jobs WHERE response_id = :rid", {"rid": response_id}
    )
    if existing:
        await db.execute(
            "UPDATE audio_jobs SET status = 'UPLOADING', wav_path = :p, updated_at = NOW() WHERE id = :id",
            {"p": upload_path, "id": existing["id"]},
        )
    else:
        await db.execute(
            "INSERT INTO audio_jobs (id, response_id, status, wav_path) VALUES (:id, :rid, 'UPLOADING', :p)",
            {"id": new_id(), "rid": response_id, "p": upload_path},
        )

    return _ok({"uploadId": upload_id, "uploadPath": upload_path, "uploadUrl": upload_url})


@router.post("/public/audio/upload/{upload_id}", status_code=201)
async def upload_audio(
    upload_id: str,
    file: UploadFile = File(...),
    responseId: str = Form(...),
    clientQcResult: str = Form(None),
    db: Database = Depends(get_db),
):
    return await _do_upload(upload_id, file, responseId, clientQcResult, db)
